### ISSUE TITLE
[3.x.x dev] Re-look at system module

### ISSUE THREAD

**Issue by aurtzy** [aurtzy/linux-autosetup#43]
Still sort of uncomfortable with current iteration of file permission-checking operations. Preferably would like to find a method that does not cause side effects - currently, open('w', ...) ... can cause a 0-byte file to be written, which isn't exactly pleasant.

- One simple solution would be to provide running as su as an option: run() could have one more argument indicating whether to run it with su cmd, that can be set in user config.

- Push harder with shell cmds? For something like read, we can do: check normal user read -> sudo check normal user read if previous failed. If both failed, then nothing will be executed and the file is assumed to not exist. This seems to have the least downsides, however one limitation (which was still a problem before) is that commands are run in a shell, meaning that the entire shell has to be run as root or not at all. Could be confusing for the user with a more dynamic method, but overall these seems better at least.

- Introducing a "root" option may actually be a decent solution. One of the main goals with these features is to make it clear what is happening and not have some weird "under the hood" things that might have side effects where the user might not realize what is happening.
As an option, "root" can be applied to the basic pack module and be available in cmd_presets, which makes it (at least moreso than before) easy to define if the preset would be put in a root shell.
The archives pack module can add onto this functionality with a dynamic system that could check whether the files would need root access and set that "root" option to true.

**Comment by aurtzy** [comment]
Potential solution?
If I instead go for the other route of running commands - as in, not running things in the shell - the "root" option becomes more viable as it focuses on a per-command basis. Using it with shell-driven commands would mean all commands are then run as root which is most likely not necessary and can cause unintended side effects.

Issues that would have to be addressed, however:
- [x] Piping. Would have been very easy before in a shell, but this might be a problem or at least more difficult to work with. A solution off the top of my head is to expand the cmd_preset and basic pack module for a "piped_cmd" and "pipe" values.
  
  Solved (?)! One downside that I'm remembering and realizing again now though is some output will not be printed to the shell when the pipe is active; however, if this is still a problem it can just be made clear that user input should not be expected when using pipes.
- [ ] Arguments. One of the main problems I faced before trying to implement this in proto-prototypes was how it would work with a command that required certain arguments in certain spots like after each option. This rules out the "breaking up into tokens" idea and leans more towards having to be more verbose with a list (for requiring positional-ness), unless I figure out maybe searching for certain expressions that indicate arguments? Sounds potentially hack-y
  
  _**Potential solution**_: use of blank spaces. Instead of some placeholder variable, the cmd list would actually have some None elements, which would be recognized as variables to replace. This means that users would have to be conscious of order (unless some standard is set in place, maybe?), but decent idea nevertheless, I'd say.
  Revision: use of any value considered False in Python could be allowed, e.g. None, empty string, 0, false.

I think these issues also mainly come from the current implementation of the files (to-be-named archives) pack module. Some thoughts on possible reworks as well:
- The Path class may actually be able to merge completely into the archives pack module, due to the uniqueness of its purpose and the customizability it still has to retain. It makes more sense for something to build off of the pack module in order to use the functionalities.
- Merging the Path class into this pack module would I think make using the pack module more sensible. paths could also have a "root" option that would indicate whether certain move/copy/etc. operations would need root permissions.
